ams/server/src/main/java/com/netease/arctic/server/table/executor/SnapshotsExpiringExecutor.java

```diff
@@ -258,31 +255,26 @@ public static void expireSnapshots(
     LOG.info("to delete {} files, success delete {} files", toDeleteFiles.get(), deleteFiles.get());
   }
 
-  public static Snapshot getClosestExpireSnapshot(UnkeyedTable changeTable, long ttl) {
-    if (changeTable.snapshots() == null) {
-      return null;
-    }
-    return Streams.stream(changeTable.snapshots())
-        .filter(snapshot -> snapshot.timestampMillis() <= ttl)
-        .min(Comparator.comparingLong(snapshot -> Math.abs(ttl - snapshot.timestampMillis())))
-        .orElse(null);
-  }
+  public static List<IcebergFileEntry> getExpiredDataFileEntries(UnkeyedTable changeTable, long ttlPoint) {
+    TableEntriesScan entriesScan = TableEntriesScan.builder(changeTable)
+        .includeFileContent(FileContent.DATA)
+        .build();
+    List<IcebergFileEntry> changeTTLFileEntries = new ArrayList<>();
 
-  public static List<DataFile> getClosestExpireDataFiles(UnkeyedTable changeTable, Snapshot closestExpireSnapshot) {
-    List<DataFile> changeTTLDataFiles = new ArrayList<>();
-    long recentExpireSnapshotId = closestExpireSnapshot.snapshotId();
-    try (CloseableIterable<FileScanTask> fileScanTasks = changeTable.newScan()
-        .useSnapshot(recentExpireSnapshotId)
-        .planFiles()) {
-      fileScanTasks.forEach(fileScanTask -> changeTTLDataFiles.add(fileScanTask.file()));
+    try (CloseableIterable<IcebergFileEntry> entries = entriesScan.entries()) {
+      entries.forEach(entry -> {
+        if (changeTable.snapshot(entry.getSnapshotId()).timestampMillis() <= ttlPoint) {
```

Comment:
It brings NPE if the snapshot is expired. We should also collect the file to `changeTTLFileEntries` if its snapshot does not exist.